docs/specifications/tests/DNSSEC-TP/dnssec16.md

```diff
@@ -0,0 +1,220 @@
+          go to next name server IP.
+       4. Else, if the DNS response contains at least one DNSKEY
+          record in the answer section, then add the name server IP and
+          the DNSKEY RRset from the answer section to the 
+          *DNSKEY RRsets* set. Also include any associated RRSIG records.
+    3. Go to next name server IP.
+
+6.  If the *CDS RRsets* set is empty then terminate this test case.
+
+7.  For each name server IP in the *CDS RRsets* set do:
+
+    1. If the CDS RRset is non-empty, get the RRset and its RRSIG
+       records, else go to next name server IP address.
+    2. If any CDS record is a "delete" CDS, then do:
+       1. If there is more than a single CDS record then add the name
+          server IP to the *Mixed Delete CDS* set.
+       2. Else, add the name server IP address to the *Delete CDS*
+          set.
+       3. Go to next name server IP.
+    3. Get the DNSKEY and its RRSIG records, if any, from the
+       *DNSKEY RRsets* for the same name server IP.
+    4. If there are no DNSKEY records, then do:
+       1. Add name server IP address to the *No DNSKEY RRset* set.
+       2. Go to next name server IP.
+    5. Repeat the following steps for each CDS record:
+       1. Compare the key tag from the CDS record with the calculated
+          key tags for the DNSKEY records.
+       2. If the CDS record does not match any DNSKEY record then add
+          the name server IP address and key tag to the
+          *No Match CDS With DNSKEY* set.
+       3. Else, if there is no RRSIG for the DNSKEY RRset created by
+          the DNSKEY record that the CDS record points at then add the
+          name server IP address and key tag of CDS record to the
+          *DNSKEY Not Signed By CDS* set.
+    6. If there are no RRSIG records for the CDS RRset, then add the
+       name server IP address to the *CDS Not Signed* set.
+    7. Else, for each RRSIG (CDS) do:
+       1. If the key tag of the RRSIG does not match any DNSKEY then
+          add the name server IP address and key tag to the
+          *CDS Signed By Unknown DNSKEY* set.
+       2. Else, if the RRSIG cannot be validated by the DNSKEY it
+          refers to by key tag, then add the name server IP and RRSIG
+          key tag to the *CDS Invalid RRSIG* set.
+    8. Go to next name server IP address.
+
+8.  If the *No DNSKEY RRset* set is non-empty, then output
+    *[DS16_CDS_WITHOUT_DNSKEY]* with all name server IP addresses
+    in the set.
+
+9.  If the *Mixed Delete CDS* set is
+    non-empty, then output *[DS16_MIXED_DELETE_CDS]* with all
+    name server IP addresses in the set.
+
+10. If the *Delete CDS* set is non-empty, then output
+    *[DS16_DELETE_CDS]* with all name server IP addresses.
+
+11. If the *No Match CDS With DNSKEY* set is non-empty then for each
+    key tag in the set output *[DS16_CDS_MATCHES_NO_DNSKEY]* with the
+    CDS key tag and the name server IP addresses in the set per key
+    tag.
```

Comment:
This could be simplified without changing the behavior, right? It'd be shorter, less complex and easier to parse.
`If the *No Match CDS With DNSKEY* set is non-empty then for each key tag in the set output`
->
`For each key tag in the *No Match CDS With DNSKEY* output`

Also, first it talks about "the key tag", then "the CDS key tag" and then "the key tag" again. Those three are all the same key tag, right?

What's going on with the tag arguments? It looks pretty complicated. Is the problem that you add them per name server IP but here you want to get them per key tag?

This goes for steps 12 and 15 as well.